### ISSUE TITLE
KeeShare 

### ISSUE THREAD

**Issue by silver7298** [strongbox-password-safe/Strongbox#393]
Would it be possible to integrate Database Sharing with KeeShare, like KeePassXC

**Comment by strongbox-mark** [comment]
Could you describe in a little more detail what aspects of KeeShare you like to see supported? 

Import or Export?

How would certificates be managed?

**Comment by silver7298** [comment]
both, import and export, or synchronization. 

I tested it with two databases and it worked very well. 

Both databases ran under KeePassXC

**Comment by strongbox-mark** [comment]
And you would want this for iOS or Mac?

**Comment by silver7298** [comment]
I wont for iOS

**Comment by basbebe** [comment]
+1

It makes it more easy to have a more granulated access policy.  
e.g. have a main keepassxc database and share only a keeshare file w/ e.g. web logins to share (e.g. via syncthing) with Strongbox on iOS.

**Comment by basbebe** [comment]
Here's another, more detailed, usecase:  
https://www.tpidg.us/keeshare-secure-password-sharing-nsc-78/

**Comment by strongbox-mark** [comment]
I think the best way to cover this functionality is just to include it under #338 - I believe KeePassXC have moved away from the certificate part of KeeShare already. The separate "share" file that is part of KeeShare would be difficult to manage under a mobile (and multi storage) environment.

I think the simplest solution is just to allow users to copy a subset of items from one database to another, which is essentially the same functionality as KeeShare. So I'm going to close this one out. I hope that makes sense.

**Comment by alexmo1997** [comment]
I'm not sure if the feature you speak about can fully replace keeshare.
I have a personal database and a work database, where I use the latter on my ipad only (and don't want to have my personal passwords there)
With keeshare, I could add the password to any of the databases, and if I put it in the keeshare group, (with the keeshare file being shared), the password would be automatically shared and even automatically updated if I change it.
Manually copying the entries on ios is thus not possible for me and I would have to open my work database separately on my desktop every time I add a new password I also need for work and copy it.

And even with more effort, it would still be worse as it is prone to get out of sync as the entries are not updated automatically.

**Comment by strongbox-mark** [comment]
Hi there, I understand this is something some users want and can see the value. However, it's a rather difficult feature to implement technically, so I'd urge you to consider other options in lieu of building this feature out. 

KeePass wasn't really designed for this kind of cross database credential sharing originally. KeePassXC have done a nice job on KeeShare, but it's a harder problem on mobile just because we don't have a simple file system to monitor and the ability to read/write arbitrary files. This is limited on iOS, and also tougher because people have their databases hosted in various clouds or SFTP/WebDAV servers. It's not impossible, just difficult.

Could you use a "Shared" database alongside your "Personal" and "Work" databases instead? Would it just be too inconvenient?

Where do you host your database and ".share" file at the moment?

**Comment by alexmo1997** [comment]
> Could you use a "Shared" database alongside your "Personal" and "Work" databases instead? Would it just be too inconvenient?

Well, it is basically what I'm doing now, and but it's pretty cumbersome, especially on my private PC where I always have to unlock, lookup stuff in and so on in two different databases.

> Where do you host your database and ".share" file at the moment?
Well, I have them both on webdav (nextcloud), on my computer as well as on my ipad.

**Comment by strongbox-mark** [comment]
Thanks for the extra detail @alexmo1997 - Will try to think about a good way to manage this.

**Comment by chriswayg** [comment]
KeeShare is an essential feature for us as well. We all transferred to KeePass from multiple years of 1Password usage, as finally the iOS and Android clients are on a similar level of quality and convenience, while 1Password is starting to abandon local vaults. The only feature we are missing is sharing a Group, a folder of credentials. 

_KeePassXC_'s  KeeShare does folder/group sharing and syncing about as well as 1Password, but it is limited to KeePassXC desktop clients. _KeePass2Android_ uses a similar, but different approach based on KeeAutoExec which  automatically opens a child database (using stored credentials) when opening the parent database.

I would prefer the KeeShare protocol, as it is already integrated into the most popular desktop cross platform application. 

I think this is certainly an **_essential feature_** for a "personal" password manager, as family sharing is very common and is also implemented well in Bitwarden, for example.

**Comment by chriswayg** [comment]
@strongbox-mark 
> Could you use a "Shared" database alongside your "Personal" and "Work" databases instead? Would it just be too inconvenient?

Mark, currently we are doing this as well. But it is  a cumbersome, almost unworkable workaround. It really is nothing like the family sharing features that we have come to expect as a standard on a "personal" password manager:

- Non-technical users cannot handle multiple databases on the same device. The whole KeePass environment is still a lot more technical than something like Bitwarden or 1Password.

- All shared credentials should ideally only exist once, as updating credentials regularly can become very cumbersome. Currently, we use the shared database more like a glorified clipboard. Which means the shared credential will reside in the main database of user A, B, and C and at least temporarily on the shared database.

- Copying items between databases is currently possible in macOS Strongbox and on KeePassXC, but not on Strongbox iOS, greatly limiting the ability to quickly share a credential when two of us only have an iPhone with us.

- Users who already use KeeShare with KeePassXC on the desktop will be reluctant to use your iOS client as Strongbox will quickly mess up the credentials in KeeShared group folder, if the user starts editing them with the expectation of things just working. **No warning is given by Strongbox (or any other client AFAIK) that KeeShare credentials will quickly be in a mess of  synced and unsynced changes.** (see my tests in the next comment)

**Comment by chriswayg** [comment]
Currently KeeShare groups are shown and are initially synchronized, but data will quickly get out of sync in various ways, if you actually try to edit or create items inside the shared group from within Strongbox. Here is what I tested:

### KeeShare setting
- Synchronize on all

### User A
- iPhone user A using KeePassium
- iPhone user A using Strongbox
- macOS user A using KeePassXC

### User C
- iPhone user C using KeePassium
- iPhone user C using Strongbox
- macOS user C using KeePassXC

### Test sync behavior

- All devices synced

- Create new item in shared Group by macOS user A using KeePassXC
	- synced correctly to...
		- iPhone user A using KeePassium
		- iPhone user A using Strongbox
		- macOS user C using KeePassXC
		- iPhone user C using Strongbox
		- iPhone user C using KeePassium

- Create a new item in shared Group by macOS user A using KeePassium
	- does not show up for...
		- iPhone user C using KeePassium
		- iPhone user C using Strongbox
		- macOS user C using KeePassXC
		
Basically credentials become private to the KeePassium user who created them in the shared group. Editing existing credentials inside the shared group creates quite a mess rather quickly... -  This makes using the KeeShare feature on KeePassXC unusable, if you are also using any KeePass apps (such as Strongbox) on devices that do not yet support KeeShare.

**_At a minimum I would expect Strongbox (and other clients) to give a clear warning when it encounters a KeePassXC database that uses a KeeShare group_**

**Comment by strongbox-mark** [comment]
Hi @chriswayg 

I hear your concerns, they're valid and I can see why this is useful. However as mentioned above, this is quite a bit more difficult to manage in the iOS environment, and makes this a large dev task for us. 

Regarding your comments around KeeShare and warnings, I will try to add this to our backlog to give warnings. 

Also, just to be straight with you, I'd like to add that we currently don't have the resources to undertake this dev effort in the next 6-12 months, possibly longer. So, it's advisable to take this under consideration and plan your setup accordingly.

Lastly, to one of your comments, which may make using the shared database idea more manageable:

> Copying items between databases is currently possible in macOS Strongbox and on KeePassXC, but not on Strongbox iOS, greatly limiting the ability to quickly share a credential when two of us only have an iPhone with us.

You can do this in Strongbox on iOS, and the copy is smart, in that it detects if you've previously copied the items to the shared database and updates them rather than creating them a new. It should be very convenient. Perhaps you hadn't seen this.

**Comment by chriswayg** [comment]
Hi @strongbox-mark,

Thanks for your response. Well, we will have to use separate databases for sharing then, as it seems there is no KeeShare or family sharing type functionality on the horizon fro Strongbox. I still hope that you would consider making family sharing a priority for development. I do not know, if the KeeShare or the KeeAutoExec approach are the best solutions to this problem. KeeShare on the desktop as implemented by KeePassXC is certainly not as user friendly as 1Password or Bitwarden. It creates a lot of moving and interdependent parts with shared databases on various shared cloud folders and multiple choices for sharing _(Import, Export, Synchronize)_, creating a multitude of ways that things could go wrong. I suppose its OK, for free Open Source software like KeePassXC, but for a (mostly) paid application like Strongbox with users expecting support, this could become a headache. 

Ideally the next version of KeePass .kdbx (version 5?) would be designed with sharing in mind, which would necessitate changes to the database structure. Maybe this could be suggested to the upstream development team.

> You can do this in Strongbox on iOS, and the copy is smart, in that it detects if you've previously copied the items to the shared database and updates them rather than creating them a new. It should be very convenient. Perhaps you hadn't seen this.

It took me a while to find out that copying a credential from one KeePass database to another on iOS is not actually using the Copy, Move or Duplicate command on the menu, but the **_Export Item_** command. Good to know, that it also uses the smart copy feature as in the desktop version.

**Comment by georgesnow** [comment]
#583 that’s sort of where this idea came from. Rather then it be database driven. This maintains compatibility across other platforms. The application presents it in way where it looks unified. But using similar mechanism like keeautoexec where it is just unlocking multiple databases and aggregating them together.

Obviously there is workflow to address. When adding a new entry there would need to be prompt most likely that asks what database do you want to save to(one or all etc..)?

Just a thought at addressing a friction point I encounter myself regularly.

**Comment by stratacast** [comment]
Since this is pretty big, would it be possible to have a fund of sorts for this specific feature?
